File: core/timeutil.py
#!/usr/bin/env python3
"""
定时任务用到的纯时间工具，不依赖项目内其它模块（settings / logger 都要 import 它，不能有环）。
"""
import re
from datetime import datetime, time, timedelta, timezone
from typing import List, Optional, Set, Tuple

SCHEDULE_OFF_VALUES = {'off', 'none', 'disabled', 'false', '0', '-'}
_TIME_PATTERN = re.compile(r'(\d{1,2}):(\d{2})')

# 周计划里的星期写法，1=周一 … 7=周日（与 isoweekday 一致）
WEEKDAY_NAMES = {
    'mon': 1, 'monday': 1, '周一': 1, '一': 1,
    'tue': 2, 'tues': 2, 'tuesday': 2, '周二': 2, '二': 2,
    'wed': 3, 'wednesday': 3, '周三': 3, '三': 3,
    'thu': 4, 'thur': 4, 'thurs': 4, 'thursday': 4, '周四': 4, '四': 4,
    'fri': 5, 'friday': 5, '周五': 5, '五': 5,
    'sat': 6, 'saturday': 6, '周六': 6, '六': 6,
    'sun': 7, 'sunday': 7, '周日': 7, '周天': 7, '日': 7, '天': 7,
}


def parse_daily_times(text: Optional[str]) -> List[time]:
    """把 ``"09:00,15:30"`` 解析成去重排序后的 time 列表；``off`` / ``none`` 等表示关闭。

    格式不对时抛 ValueError，配合 settings 的校验在启动时直接报错。
    """
    if text is None:
        return []
    raw = text.strip()
    if not raw or raw.lower() in SCHEDULE_OFF_VALUES:
        return []

    result = set()
    for part in raw.split(','):
        part = part.strip()
        if not part:
            continue
        matched = _TIME_PATTERN.fullmatch(part)
        if not matched:
            raise ValueError(f'时刻格式错误: {part!r}，应为 HH:MM，多个时刻用逗号分隔')
        hour, minute = int(matched.group(1)), int(matched.group(2))
        if not (0 <= hour <= 23 and 0 <= minute <= 59):
            raise ValueError(f'时刻越界: {part!r}')
        result.add(time(hour, minute))
    return sorted(result)


def parse_weekdays(text: Optional[str]) -> Set[int]:
    """把 ``"Mon,Thu"`` / ``"周一"`` 解析成 {1, 4}；空表示不限定星期"""
    if not text or not text.strip():
        return set()
    result = set()
    for part in text.replace('、', ',').split(','):
        part = part.strip().lower()
        if not part:
            continue
        if part.isdigit() and 1 <= int(part) <= 7:
            result.add(int(part))
            continue
        if part not in WEEKDAY_NAMES:
            raise ValueError(f'星期格式错误: {part!r}，可写 Mon / 周一 / 1')
        result.add(WEEKDAY_NAMES[part])
    return result


def parse_weekly_schedule(text: Optional[str]) -> Tuple[Set[int], List[time]]:
    """把 ``"Mon 09:00"`` 解析成 ({1}, [09:00])；省略星期表示每天，``off`` 表示关闭。

    星期与时刻之间用空格分隔，各自都能用逗号写多个，例如 ``"Mon,Thu 09:00,18:00"``。
    """
    if text is None:
        return set(), []
    raw = text.strip()
    if not raw or raw.lower() in SCHEDULE_OFF_VALUES:
        return set(), []

    parts = raw.split(None, 1)
    if len(parts) == 1:
        return set(), parse_daily_times(parts[0])
    return parse_weekdays(parts[0]), parse_daily_times(parts[1])


def describe_daily_times(times: List[time], weekdays: Optional[Set[int]] = None) -> str:
    if not times:
        return '已关闭'
    clock = ' / '.join(t.strftime('%H:%M') for t in sorted(times))
    if not weekdays:
        return f'每天 {clock}'
    labels = '、'.join('一二三四五六日'[day - 1] for day in sorted(weekdays))
    return f'每周{labels} {clock}'


def next_daily_occurrence(now: datetime, times: List[time], weekdays: Optional[Set[int]] = None) -> datetime:
    """下一个到点时刻。weekdays 非空时只在这些星期触发。now 与返回值同为本地时间。"""
    ordered = sorted(times)
    for offset in range(8):  # 最多找一周，一定能落到某一天
        day = now.date() + timedelta(days=offset)
        if weekdays and day.isoweekday() not in weekdays:
            continue
        for t in ordered:
            candidate = datetime.combine(day, t, tzinfo=now.tzinfo)
            if candidate > now:
                return candidate
    raise ValueError('无法计算下次运行时间')


def local_now() -> datetime:
    """带时区的本地当前时间（容器里由 TZ 决定）"""
    return datetime.now().astimezone()


def to_utc_iso(value: Optional[datetime]) -> Optional[str]:
    """datetime → UTC ISO 字符串（Z 结尾）；naive 值按本地时间理解"""
    if value is None:
        return None
    if value.tzinfo is None:
        value = value.astimezone()
    return value.astimezone(timezone.utc).isoformat().replace('+00:00', 'Z')

File: internal/timeutil/timeutil.go
// Package timeutil provides the package implementation.
//
// Implementation note.
// Implementation note.
package timeutil

import (
	"fmt"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"time"
)

// Implementation note.
type ClockTime struct {
	Hour   int
	Minute int
}

func (c ClockTime) String() string { return fmt.Sprintf("%02d:%02d", c.Hour, c.Minute) }

func (c ClockTime) before(other ClockTime) bool {
	return c.Hour < other.Hour || (c.Hour == other.Hour && c.Minute < other.Minute)
}

// Implementation note.
var offValues = map[string]bool{
	"off": true, "none": true, "disabled": true, "false": true, "0": true, "-": true,
}

var timePattern = regexp.MustCompile(`^(\d{1,2}):(\d{2})$`)

// Implementation note.
var weekdayNames = map[string]int{
	"mon": 1, "monday": 1, "周一": 1, "一": 1,
	"tue": 2, "tues": 2, "tuesday": 2, "周二": 2, "二": 2,
	"wed": 3, "wednesday": 3, "周三": 3, "三": 3,
	"thu": 4, "thur": 4, "thurs": 4, "thursday": 4, "周四": 4, "四": 4,
	"fri": 5, "friday": 5, "周五": 5, "五": 5,
	"sat": 6, "saturday": 6, "周六": 6, "六": 6,
	"sun": 7, "sunday": 7, "周日": 7, "周天": 7, "日": 7, "天": 7,
}

// Implementation note.
// Implementation note.
func ParseDailyTimes(text string) ([]ClockTime, error) {
	raw := strings.TrimSpace(text)
	if raw == "" || offValues[strings.ToLower(raw)] {
		return nil, nil
	}

	seen := make(map[ClockTime]bool)
	var result []ClockTime
	for _, part := range strings.Split(raw, ",") {
		part = strings.TrimSpace(part)
		if part == "" {
			continue
		}
		matched := timePattern.FindStringSubmatch(part)
		if matched == nil {
			return nil, fmt.Errorf("invalid time format %q; use HH:MM and separate multiple times with commas", part)
		}
		hour, _ := strconv.Atoi(matched[1])
		minute, _ := strconv.Atoi(matched[2])
		if hour < 0 || hour > 23 || minute < 0 || minute > 59 {
			return nil, fmt.Errorf("time is out of range: %q", part)
		}
		c := ClockTime{Hour: hour, Minute: minute}
		if !seen[c] {
			seen[c] = true
			result = append(result, c)
		}
	}
	sort.Slice(result, func(i, j int) bool { return result[i].before(result[j]) })
	return result, nil
}

// Implementation note.
func ParseWeekdays(text string) (map[int]bool, error) {
	if strings.TrimSpace(text) == "" {
		return nil, nil
	}
	result := make(map[int]bool)
	for _, part := range strings.Split(strings.ReplaceAll(text, "、", ","), ",") {
		part = strings.ToLower(strings.TrimSpace(part))
		if part == "" {
			continue
		}
		if n, err := strconv.Atoi(part); err == nil {
			if n < 1 || n > 7 {
				return nil, fmt.Errorf("invalid weekday %q; use Mon or an ISO weekday number", part)
			}
			result[n] = true
			continue
		}
		day, ok := weekdayNames[part]
		if !ok {
			return nil, fmt.Errorf("invalid weekday %q; use Mon or an ISO weekday number", part)
		}
		result[day] = true
	}
	return result, nil
}

// Implementation note.
// Implementation note.
func ParseWeeklySchedule(text string) (map[int]bool, []ClockTime, error) {
	raw := strings.TrimSpace(text)
	if raw == "" || offValues[strings.ToLower(raw)] {
		return nil, nil, nil
	}
	parts := strings.SplitN(raw, " ", 2)
	if len(parts) == 1 {
		times, err := ParseDailyTimes(parts[0])
		return nil, times, err
	}
	weekdays, err := ParseWeekdays(parts[0])
	if err != nil {
		return nil, nil, err
	}
	times, err := ParseDailyTimes(strings.TrimSpace(parts[1]))
	return weekdays, times, err
}

var weekdayLabels = []string{"Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"}

// Implementation note.
func Describe(times []ClockTime, weekdays map[int]bool) string {
	if len(times) == 0 {
		return "Disabled"
	}
	clocks := make([]string, len(times))
	for i, t := range times {
		clocks[i] = t.String()
	}
	clock := strings.Join(clocks, " / ")
	if len(weekdays) == 0 {
		return "Daily " + clock
	}
	days := make([]int, 0, len(weekdays))
	for d := range weekdays {
		days = append(days, d)
	}
	sort.Ints(days)
	labels := make([]string, len(days))
	for i, d := range days {
		labels[i] = weekdayLabels[d-1]
	}
	return "Weekly " + strings.Join(labels, ", ") + " " + clock
}

// Implementation note.
func ISOWeekday(t time.Time) int {
	if w := int(t.Weekday()); w == 0 {
		return 7
	} else {
		return w
	}
}

// Implementation note.
// Implementation note.
func NextOccurrence(now time.Time, times []ClockTime, weekdays map[int]bool) (time.Time, bool) {
	if len(times) == 0 {
		return time.Time{}, false
	}
	ordered := append([]ClockTime(nil), times...)
	sort.Slice(ordered, func(i, j int) bool { return ordered[i].before(ordered[j]) })

	for offset := 0; offset < 8; offset++ { // Search at most one week; a matching day must exist.
		day := now.AddDate(0, 0, offset)
		if len(weekdays) > 0 && !weekdays[ISOWeekday(day)] {
			continue
		}
		for _, c := range ordered {
			candidate := time.Date(day.Year(), day.Month(), day.Day(), c.Hour, c.Minute, 0, 0, now.Location())
			if candidate.After(now) {
				return candidate, true
			}
		}
	}
	return time.Time{}, false
}

// Implementation note.
func UTCISO(t time.Time) *string {
	if t.IsZero() {
		return nil
	}
	s := t.UTC().Format("2006-01-02T15:04:05.999999Z")
	return &s
}

// Implementation note.
func NowISO() string { return time.Now().UTC().Format("2006-01-02T15:04:05.999999Z") }
